import asyncio
import logging
from concurrent import futures
import json
import uuid

import grpc

# å¯¼å…¥æˆ‘ä»¬ç”Ÿæˆçš„gRPCä»£ç 
import intelligence_pb2
import intelligence_pb2_grpc

# --- gRPC Service Implementation ---
class IntelligenceServiceImpl(intelligence_pb2_grpc.IntelligenceServiceServicer):
    def InterpretText(self, request: intelligence_pb2.InterpretRequest, context):
        logging.info(f"Received text: '{request.user_text}' from session '{request.session_id}'")

        # æ¨¡æ‹Ÿ AI å“åº”é€»è¾‘
        user_text = request.user_text.lower()
        
        # æ£€æŸ¥æ˜¯å¦åŒ…å«ç”µè¯å·ç æ›´æ–°ç›¸å…³çš„å…³é”®è¯
        if any(keyword in user_text for keyword in ['ç”µè¯', 'æ‰‹æœº', 'å·ç ', 'phone', 'update']):
            # æ¨¡æ‹Ÿæ›´æ–°ç”µè¯å·ç çš„æ„å›¾
            mock_response = {
                "employee_id": str(uuid.uuid4()),
                "new_phone_number": "13800138000"
            }
            
            return intelligence_pb2.InterpretResponse(
                intent="update_phone_number",
                structured_data_json=json.dumps(mock_response)
            )
        
        # æ£€æŸ¥æ˜¯å¦åŒ…å«æŸ¥è¯¢å‘˜å·¥ç›¸å…³çš„å…³é”®è¯
        elif any(keyword in user_text for keyword in ['å‘˜å·¥', 'æŸ¥è¯¢', 'æŸ¥æ‰¾', 'employee', 'search']):
            # æ¨¡æ‹ŸæŸ¥è¯¢å‘˜å·¥çš„æ„å›¾
            mock_response = {
                "employee_id": str(uuid.uuid4())
            }
            
            return intelligence_pb2.InterpretResponse(
                intent="get_employee_manager",
                structured_data_json=json.dumps(mock_response)
            )
        
        # é»˜è®¤è¿”å›žæ— æ„å›¾æ£€æµ‹
        return intelligence_pb2.InterpretResponse(
            intent="no_intent_detected", 
            structured_data_json="{}"
        )

async def serve() -> None:
    port = "50051"
    server = grpc.aio.server(futures.ThreadPoolExecutor(max_workers=10))
    intelligence_pb2_grpc.add_IntelligenceServiceServicer_to_server(
        IntelligenceServiceImpl(), server
    )
    server.add_insecure_port(f"[::]:{port}")
    logging.info(f"ðŸ§™ Python AI Service 'The Wizard Tower' (MOCK) is listening on gRPC port {port}")
    await server.start()
    await server.wait_for_termination()

if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    asyncio.run(serve()) 